### PR TITLE
Adjustments for Objective-C, Android and general builds.

### PR DESCRIPTION
**Pull request by j-jorge**

### CODE REVIEW

**Comment by dcourtois** [comment]
Hi,

Thanks for the pull request. I'll take a closer look at it as soon as I have some time, but after a quick look, I have a question : as I understand, in your last commit, only the generated compilation units are in the project now ? If I understood that correctly, I need this to be an option (preferably a configuration option) because I need the original  files to still be included in the project and appear in my IDE.

Also, those files have their compilation disabled, so they shouldn't cause any problem. If they do, maybe that's a problem with the "generator" that you're using which does not support per-file compilation disabling ?

**Comment by dcourtois** [comment]
Hi,

Any news ?

Also I have another question: I don't really see the point of the second commit. I mean, if the addon generates an empty compilation unit, it means you enabled compilation units on a project with less files than the number of CPU cores ... which kind of defeat the purpose of this addon :/

And even if it does generate an empty file, does it create an error ? (I'm not sure how compilers behave when trying to compile an empty file)

**Comment by j-jorge** [comment]
Hi,

Without the last commit all files, both original and units, are listed in the Android makefiles generated by this [Premake NDK module](https://github.com/IsCoolEntertainment/premake-ndk) (this is a fork of [this quite old module](https://bitbucket.org/secint/premake-ndk/src/default/)). If I remember well the same problem occurs with the XCode projects generated by Premake. If you know how to filter out the original files I can update upstream's code to not list them.

As for the second commit, compiling empty files and linking them in a static library generates `<file> has no symbols` warnings in ranlib on OSX/iOS. Even if its a non-blocking message I think it is better not to have it. It happens that I had empty compilation units by setting like 8 units at the root of my project then having some subprojects compiling less than 8 files under some circumstances. Like when targeting Linux or Android there was less than 8 files but when targeting iOS there was more. I thought it would be more clear to not generate the file in the `compilationunit` module rather than trying to figure out if there would be less files than expected on the project's side.

**Comment by dcourtois** [comment]
Hi again, thanks for taking the time to answer :)

Ok, for the empty compilation files: even though IMO tools shouldn't have to worry about those kind of details, and it should be the responsibility of each project to configure it as needed, the change is small enough, so I'm fine with it.

But for the last commit, I can't accept the PR as it is: up until now, generating a compilation unit enabled project on VS (for instance) generates a project where you have all files in the usual tree view, allowing you to work on the project in the IDE, navigate your sources, etc. This commit changes this behavior for everyone, every generator.

As I said: you'll need to make this behavior (removing sources that are included in compilation units from the project) a per-config option to not force this change on everyone (and the default should respect the current behavior)

Also note that the addon disables compilation of files that are already included in a compilation unit. If the original files appear on the final makefile, it means that the generator you're using (NDK module in your case) does not correctly handle per-file configurations.

And finally, I added some notes on the commits (yeah I know I'm being picky :p)

File: _preload.lua [hunk 1]
```diff
@@ -36,6 +36,14 @@ premake.api.register {
        tokens = true
 }
 
+--
+-- Default unit extension
```
Comment (dcourtois) [inline _preload.lua:40]: A little more doc ? As it is, I have to read the source code where it's used to understand how it works. Maybe something like:

```lua
-- Compilation unit extensions.
--
-- By default, either .c or .cpp extension are used for generated compilation units.
-- But you can override this extension per-language to let it handle objective-C or
-- any other.
--
-- Example use
-- ```
-- filter {}
-- 	compilationunitextensions {
--		"CPP" = ".cc",
--		"OBJC" = ".mm"
--	}
-- ```
```

File: compilationunit.lua [hunk 2]
```diff
@@ -247,7 +247,19 @@ end
...
 --
 function premake.extensions.compilationunit.getCompilationUnitName(cfg, index, shortName)
-	return premake.extensions.compilationunit.compilationunitname .. index .. iif(cfg.language == "C", ".c", ".cpp")
+
+	local language = cfg.language
+	local extension = nil
+	
+	if ( cfg.compilationunitextensions ~= nil ) then
+		  extension = cfg.compilationunitextensions[ language ]
+	end
```
Comment (dcourtois) [inline compilationunit.lua:259]: code convention : no parenthesis please, and no spaces in cfg.compilationunitextensions[language]
also please check whitespaces : there is a space before `extension`

File: compilationunit.lua [hunk 3]
```diff
@@ -247,7 +247,19 @@ end
...
+	
+	if extension == nil then
+		 extension = iif(language == "C", ".c", ".cpp")
+	end
```
Comment (dcourtois) [inline compilationunit.lua:263]: whitespace before `extension`

File: compilationunit.lua [hunk 4]
```diff
@@ -62,7 +62,8 @@ function premake.extensions.compilationunit.customBakeFiles(base, prj)
 		end
 
 		-- add the compilation units for premake
-		for i = 1, cu.numcompilationunits do
+                local count = math.min(#cu.compilationunits[cfg], cu.numcompilationunits)
```
Comment (dcourtois) [inline compilationunit.lua]: please add a comment before to explain what it's doing. Just a small note, something like:

```
-- add the compilation units for premake
-- note: avoid generating empty compilation units if we have less files in the project
-- than the number of compilation units.
```

**Review by dcourtois — COMMENTED** [review]
Hi,

It's been a while since my review, I just wanted to know if you're going to make the changes ?

**Comment by j-jorge** [comment]
Hi,

Well, it's a bit difficult to find time to come back on this but I will try to do it this month.

**Comment by dcourtois** [comment]
Yup, I can understand that :)
If you don't have time, I can do it. The 2 last commits are pretty easy to redo, and the first one can be done quite easily too: I just need to "optionize" it. For that, I have 2 options:
1. make it an option, e.g. something like `premake5 --compilationunit=8 --compilationunitsonly=true ...`
2. make it an API command to make it work per-config directly in the projects' .lua scripts.

I personally think the first is better (integrated doc through `--help`, simpler code, no change in your lua scripts), but since I won't be using it, I'd like to know what you think ?

**Comment by j-jorge** [comment]
I agree with you on the `--compilationunitsonly` option, it is clearly a project-wide setting, I see no reason to have it at the config level.

**Comment by dcourtois** [comment]
Nice ! And now, a last question: should I wait for your fixes, or are you ok with me closing this PR and making the modifications by myself ? :)

**Comment by j-jorge** [comment]
I'm OK if you do the modifications by yourself. Thanks :)

**Comment by j-jorge** [comment]
Hi,
I've made the requested changes in the `fix-upstream-review` branch in our fork. I'm trying to figure how to update this PR without updating our master.

**Comment by j-jorge** [comment]
It seems that changing the source branch is not possible, I'm just going to open new PRs. Sorry for the noise.

**Comment by dcourtois** [comment]
Ahh ! Sorry I totally forgot about this issue ! I merge 2 of the issues and there is just some space -> tabs to change on the third (made a comment)
Thanks a lot for taking the time to work on those pull request, and apologies again (going to close this issue now that it's split into the 3 other ones)

**Pull request closed without merging**